### PR TITLE
Update ingest script to also output vector tiles

### LINKED ISSUE #34: Adapt ingestion script to output vector tiles
As a follow-up to #27, the script will need to be adapted to output vector tiles.

To be able to function correctly with selection on the frontend, the geolevel hierarchy indices should be placed within the properties of each polygon in the vector tile. An example block may have the properties: {countyIndex: 3, blockgroupIndex: 2, blockIndex: 4, ...}. This is in addition to the other data (such as demographic statistics) that should be included within the vector tile.

### PR DESCRIPTION
**Pull request by maurizi**
## Overview

Adds output of separate vector tiles file pyramids per geolevel to ingest script

### Demo

```
src/manage [ ./bin/run process-geojson ~/Downloads/DE.geojson -o output -l block,tract,county                                             feature/mvm/topojson-to-vector-tiles ] 2:55 PM 
Reading base GeoJSON: /home/mike/Downloads/DE.geojson... done                                                                                                                            
GeoJSON contains 24115 features                                                                                                                                                         
Converting to topojson with base geolevel: block                                                                                                                                        
Presimplifying using planar triangle area                                                                                                                                               
Simplifying block geounits with minWeight: 0.001                                                                                                                                        
Grouping geoLevel "tract"                                                                                                                                                               
Merging 217 features                                                                                                                                                                    
Grouping geoLevel "county"                                                                                                                                                              
Merging 3 features                                                                                                                                                                      
Writing topojson file                                                                                                                                                                   
Converting topojson to geojson for block                                                                                                                                                
Converting topojson to geojson for tract                                                                                                                                                
Converting topojson to geojson for county                                                                                                                                               
Converting geojson to vectortiles for block                                                                                                                                             
tippecanoe --force --output-to-directory=output/block --maximum-zoom=g --minimum-zoom=8 output/block.geojson                                                                            
For layer 0, using name "block"                                                                                                                                                         
24115 features, 1176099 bytes of geometry, 809120 bytes of separate metadata, 9881 bytes of string pool                                                                                 
Choosing a maxzoom of -z9 for features about 909 feet (277 meters) apart                                                                                                                
  99.9%  9/148/194                                                                                                                                                                      
Converting geojson to vectortiles for tract                                                                                                                                             
tippecanoe --force --output-to-directory=output/tract --maximum-zoom=g --minimum-zoom=0 output/tract.geojson                                                                            
For layer 0, using name "tract"                                                                                                                                                         
217 features, 82324 bytes of geometry, 5630 bytes of separate metadata, 6519 bytes of string pool                                                                                       
Choosing a maxzoom of -z5 for features about 13966 feet (4257 meters) apart                                                                                                             
Choosing a maxzoom of -z8 for resolution of about 1244 feet (379 meters) within features                                                                                                
  99.9%  8/74/97                                                                                                                                                                        
Converting geojson to vectortiles for county                                                                                                                                            
tippecanoe --force --output-to-directory=output/county --maximum-zoom=g --minimum-zoom=0 output/county.geojson                                                                          
For layer 0, using name "county"
3 features, 4410 bytes of geometry, 79 bytes of separate metadata, 201 bytes of string pool
Choosing a maxzoom of -z1 for features about 204113 feet (62214 meters) apart
Choosing a maxzoom of -z7 for resolution of about 3223 feet (982 meters) within features
  94.9%  7/37/48
Writing static data file for population
Writing static data file for white
Writing static data file for black
Writing static data file for asian
Writing static data file for hispanic
Writing static data file for other
Writing tract index file
Writing county index file
Writing static metadata file
```

```
rc/manage [ ls -R output                                                                                                               feature/mvm/topojson-to-vector-tiles * ] 2:55 PM 
output:                                                                                                                                                                                  
asian.buf  block          county      county.geojson  other.buf       static-metadata.json  tract      tract.geojson                                                                     
black.buf  block.geojson  county.buf  hispanic.buf    population.buf  topo.json             tract.buf  white.buf                                                                         
                                                                                                                                                                                         
output/block:                                                                                                                                                                            
8  9  metadata.json                                                                                                                                                                      
                                                                                                                                                                                         
output/block/8:                                                                                                                                                                          
74                                                                                                                                                                                       
                                                                                                                                                                                         
output/block/8/74:                                                                                                                                                                       
97.pbf  98.pbf                                                                                                                                                                           
                                                                                                                                                                                         
output/block/9:                                                                                                                                                                          
148  149                                                                                                                                                                                 
                                                                                                                                                                                         
output/block/9/148:                                                                                                                                                                      
194.pbf  195.pbf  196.pbf                                                                                                                                                                
                                                                                                                                                                                         
output/block/9/149:                                                                                                                                                                      
195.pbf  196.pbf                                                                                                                                                                         
                                                                                                                                                                                         
output/county:                                                                                                                                                                           
0  1  2  3  4  5  6  7  metadata.json

output/county/0:
0

output/county/0/0:
0.pbf

output/county/1:
0

output/county/1/0:
0.pbf

output/county/2:
1

output/county/2/1:
1.pbf

output/county/3:
2

output/county/3/2:
3.pbf

output/county/4:
4

utput/county/4/4:                                                                                                                                                               [20/650]
6.pbf

output/county/5:
9

output/county/5/9:
12.pbf

output/county/6:
18

output/county/6/18:
24.pbf

output/county/7:
37

output/county/7/37:
48.pbf  49.pbf

output/tract:
0  1  2  3  4  5  6  7  8  metadata.json

output/tract/0:
0

output/tract/0/0:
0.pbf

output/tract/1:
0

output/tract/1/0:
0.pbf

output/tract/2:
1

output/tract/2/1:
1.pbf

output/tract/3:
2

output/tract/3/2:
3.pbf

output/tract/4:
4

output/tract/4/4:
6.pbf

output/tract/5:
9

output/tract/5/9:
12.pbf

output/tract/6:
18

output/tract/6/18:
24.pbf

output/tract/7:
37

output/tract/7/37:
48.pbf  49.pbf

output/tract/8:
74

output/tract/8/74:
97.pbf  98.pbf
```

## Testing Instructions

You'll need `tippecanoe` on your `PATH`.
I used [the instructions to build from source](https://github.com/mapbox/tippecanoe#installation).

- `cd src/manage && yarn install`
- `./bin/run process-geojson ~/Downloads/DE.geojson -o output -l block,tract,county`
- `NODE_OPTIONS="--max-old-space-size=8192" ./bin/run process-geojson ~/Downloads/PA.geojson -o output -b`

Closes #34

### CODE REVIEW

File: src/manage/src/commands/process-geojson.ts [hunk 1]
```diff
@@ -40,6 +48,18 @@ it when necessary (file sizes ~1GB+).
       default: "block,blockgroup,county"
     }),
 
+    levelMinZoom: flags.string({
```
Comment (kshepard) [inline src/manage/src/commands/process-geojson.ts]: After adding arguments/etc. to this command, the CLI README should be regenerated by running `./node_modules/.bin/oclif-dev readme`.

File: src/manage/src/commands/process-geojson.ts [hunk 2]
```diff
@@ -64,9 +84,16 @@ it when necessary (file sizes ~1GB+).
   async run(): Promise<void> {
     const { args, flags } = this.parse(ProcessGeojson);
     const geoLevels = flags.levels.split(",");
+    const minZooms = flags.levelMinZoom.split(",");
+    const maxZooms = flags.levelMaxZoom.split(",");
     const demographics = flags.demographics.split(",");
     const simplification = parseFloat(flags.simplification);
 
+    if (geoLevels.length !== minZooms.length || geoLevels.length !== maxZooms.length) {
+      this.log(`GeoJSON must have features, exiting`);
```
Comment (kshepard) [inline src/manage/src/commands/process-geojson.ts]: Looks like a copy-pasted error message.

File: src/manage/src/commands/process-geojson.ts [hunk 3]
```diff
@@ -301,4 +332,37 @@ it when necessary (file sizes ~1GB+).
...
+    maxZooms: readonly string[]
+  ): void {
+    geoLevels.forEach((geoLevel, idx) => {
+      const minimumZoom = minZooms[idx];
+      const maximumZoom = maxZooms[idx];
+      const tileDir = join(dir, geoLevel);
+      this.log(`Converting geojson to vectortiles for ${geoLevel}`);
+      tippecanoe(
+        [join(dir, `${geoLevel}.geojson`)],
+        { force: true, outputToDirectory: tileDir, maximumZoom, minimumZoom },
```
Comment (kshepard) [inline src/manage/src/commands/process-geojson.ts]: I was expecting one set of vector tiles, as is done in the prototype here: https://github.com/azavea/districtbuilder-2/blob/486e50d974cc61d77b939b765ad1ed4de454c8b6/public/process/process.sh#L44. In that script, the `mbtiles` files are written out separately, and then joined with `tile-join`. That wasn't spelled out explicitly in the issue, so we can discuss and add a follow-up issue.

File: src/manage/src/commands/process-geojson.ts [hunk 4]
```diff
@@ -301,4 +332,37 @@ it when necessary (file sizes ~1GB+).
...
+
+  // Convert TopoJSON to GeoJSON and write to disk
+  writeIntermediaryGeoJson(
+    dir: string,
+    topology: Topology<Objects<{}>>,
+    geoLevels: readonly string[]
+  ): void {
+    geoLevels.forEach(geoLevel => {
+      this.log(`Converting topojson to geojson for ${geoLevel}`);
+      const geojson = topo2feature(topology, topology.objects[geoLevel]);
```
Comment (kshepard) [inline src/manage/src/commands/process-geojson.ts]: It doesn't look like any special handling is made to accommodate the geolevel indices properties (as discussed in #34). See the answer to the third question [here](https://github.com/PublicMapping/district-builder-2/pull/18#issuecomment-572671925) for an idea on why they're needed.

File: src/manage/src/commands/process-geojson.ts (hunk 3 above)
Comment (maurizi) [reply src/manage/src/commands/process-geojson.ts]: I vendored the `tippecanoe` library I was using and adapted it to also wrap `tile-join` so I could tackle this.

File: src/manage/src/commands/process-geojson.ts (hunk 4 above)
Comment (maurizi) [reply src/manage/src/commands/process-geojson.ts]: We now have an `index` property on all vector tile features. As we discussed, we're planning to distribute all other properties via static array buffers which we'll access via this `index`.

**Comment by kshepard** [comment]
Cool, these changes are sounding good, I'll take them for a spin. Just curious: have you attempted to load up these tiles on a map to see what they look like, just as a sanity check? If not, I think it'd be a good thing to test out that should be relatively straightforward.

**Pull request merged**